packages/network-controller/src/NetworkController.ts

```diff
@@ -92,6 +99,74 @@ export type NetworkConfiguration = {
   };
 };
 
+/**
+ * The collection of network configurations in state.
+ */
+type NetworkConfigurations = Record<
+  NetworkConfigurationId,
+  NetworkConfiguration & { id: NetworkConfigurationId }
+>;
+
+/**
+ * `Object.keys()` is intentionally generic: it returns the keys of an object,
+ * but it cannot make guarantees about the contents of that object, so the type
+ * of the keys is merely `string[]`. While this is technically accurate, it is
+ * also unnecessary if we have an object that we own and whose contents are
+ * known exactly.
+ *
+ * TODO: Move to @metamask/utils.
+ *
+ * @param object - The object.
+ * @returns The keys of an object, typed according to the type of the object
+ * itself.
+ */
+function knownKeysOf<K extends PropertyKey>(object: Partial<Record<K, any>>) {
+  return Object.keys(object) as K[];
+}
+
+/**
+ * Asserts that the given value is of the given type if the given functions
+ * returns a truthy result.
+ *
+ * @param _value - Some value.
+ * @param test - The function to run.
+ * @param message - The message to throw if the function does not return a
+ * truthy result.
+ * @throws if the function does not return a truthy result.
+ */
+function assertOfType<Type>(
+  _value: unknown,
+  test: () => boolean,
+  message: string,
+): asserts _value is Type {
+  assert.ok(test(), message);
+}
+
+/**
+ * Returns a portion of the given object with only the given keys.
```

Comment:
Nit: Maybe worth clarifying that it also ensures the keys exist

Reply:
The assertion is performed on the resulting object and is purely for type assignment purposes; in reality, an error will never be thrown, because the loop picks the requested keys from the object and places all of them in the new object.

Reply:
Ah, right... so even if a property is missing, it'll at least be set to `undefined` in the resulting object.